hedera-node/hedera-app/src/main/java/com/hedera/node/app/tss/TssLibraryImpl.java

```diff
@@ -0,0 +1,152 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hedera.node.app.tss;
+
+import static com.hedera.node.app.tss.handlers.TssUtils.SIGNATURE_SCHEMA;
+
+import com.hedera.cryptography.bls.BlsPrivateKey;
+import com.hedera.cryptography.bls.BlsPublicKey;
+import com.hedera.cryptography.bls.BlsSignature;
+import com.hedera.cryptography.tss.api.TssMessage;
+import com.hedera.cryptography.tss.api.TssParticipantDirectory;
+import com.hedera.cryptography.tss.api.TssParticipantPrivateInfo;
+import com.hedera.cryptography.tss.api.TssPrivateShare;
+import com.hedera.cryptography.tss.api.TssPublicShare;
+import com.hedera.cryptography.tss.api.TssService;
+import com.hedera.cryptography.tss.api.TssShareSignature;
+import com.hedera.cryptography.tss.impl.Groth21Service;
+import com.hedera.node.app.spi.AppContext;
+import com.hedera.node.app.tss.api.TssLibrary;
+import com.hedera.pbj.runtime.io.buffer.Bytes;
+import com.swirlds.state.lifecycle.info.NodeInfo;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.security.SecureRandom;
+import java.util.List;
+import java.util.function.Supplier;
+import javax.inject.Inject;
+
+public class TssLibraryImpl implements TssLibrary {
+    private static final BlsPrivateKey AGGREGATED_PRIVATE_KEY =
+            BlsPrivateKey.create(SIGNATURE_SCHEMA, new SecureRandom());
+    private byte[] message = new byte[0];
+
+    private TssService tssService;
+    private SecureRandom random = new SecureRandom();
+    private Supplier<NodeInfo> selfNodeInfo;
+
+    @Inject
+    public TssLibraryImpl(AppContext appContext) {
+        this.tssService = new Groth21Service(SIGNATURE_SCHEMA, random);
+        this.selfNodeInfo = appContext.selfNodeInfoSupplier();
+    }
+
+    @NonNull
+    @Override
+    public TssMessage generateTssMessage(@NonNull final TssParticipantDirectory tssParticipantDirectory) {
+        return tssService.genesisStage().generateTssMessage(tssParticipantDirectory);
+    }
+
+    @NonNull
+    @Override
+    public TssMessage generateTssMessage(
+            @NonNull TssParticipantDirectory tssParticipantDirectory, @NonNull TssPrivateShare privateShare) {
+        return tssService.rekeyStage().generateTssMessage(tssParticipantDirectory, privateShare);
+    }
+
+    @Override
+    public boolean verifyTssMessage(
+            @NonNull TssParticipantDirectory participantDirectory, @NonNull Bytes tssMessageBytes) {
+        try {
+            final var tssMessage = tssService.messageFromBytes(participantDirectory, tssMessageBytes.toByteArray());
+            return tssService.rekeyStage().verifyTssMessage(participantDirectory, null, tssMessage);
+        } catch (Exception e) {
+            return false;
+        }
+    }
```

Comment:
Parsing a message is a computationally intense operation.
The library provides two separate operations for parsing and validating to allow the reuse of the parsed message if it results in a valid one.

Reply:
talked about this in a call, will be considered in the future

Reply:
Agreed. Will make sure we are not discarding or recomputing again by going through the code.